src/model.py

```diff
@@ -2259,35 +2211,34 @@ def forward(self, batch):
                     need_weights=True
                 )
 
-                attn_inter = torch.cat((cnn_outputs[first_task].permute(0, 2, 1), relation_attn.permute(1,0,2)), dim=-1)
-                attn_logits = self.components[first_task]["attn_linear"](attn_inter) 
+                attn_inter = torch.cat((cnn_outputs[first_task].permute(0, 2, 1), relation_attn.permute(1,0,2)), dim=-1)  # embedding dim
+                attn_logits = relation_components["linear"](attn_inter)  # produce logits for first_task using info from attn
+
+                # stack attention outputs for this relation (averaged after all stacks complete)
+                if first_task not in ("shared", "all"):
+                    outputs[first_task].append(attn_logits)
 
-                # store task inputs for next stack, considering multiple relations for first task
-                if (first_task == prev_first) or (first_task == "shared") or (prev_first is None):
-                    same_task.append(attn_inter)
+                # append attn output to list if first task present, else create new list
+                if first_task in relation_inters:
+                    relation_inters[first_task].append(relation_attn.permute(1,0,2))
                 else:
-                    same_task = [attn_inter]
+                    relation_inters[first_task] = [relation_attn.permute(1,0,2)]
 
-                if first_task in task_inputs:
-                    re_encoded = self.components[first_task]["re_encode"](torch.stack(same_task).mean(dim=0))
-                    task_inputs[first_task] = re_encoded.permute(0, 2, 1)
-                prev_first = first_task    
+            # concatenate all attention outputs using same first task w/ that task's conv outputs
+            for task in relation_inters:
+                if task in self.subtasks:
+                    all_task_inters = torch.cat([cnn_outputs[task].permute(0, 2, 1)] + relation_inters[task], dim=-1)
+                    task_inputs[task] = self.components[task]["re_encode"](all_task_inters).permute(0, 2, 1)  # batch, sequence, embedding
```

Comment:
Re-encoding outputs from all attentions outputs and convolution output for these tasks.